utils/serde/src/main/java/io/apicurio/registry/utils/serde/JsonSchemaKafkaSerializer.java

```diff
@@ -0,0 +1,186 @@
+            @Override
+            protected SchemaValidator toSchema(Response response) {
+                String schema = response.readEntity(String.class);
+                return api.loadSchema(new StringSchemaSource(schema));
+            }
+        };
+
+    }
+
+    /**
+     * @see org.apache.kafka.common.serialization.Serializer#configure(java.util.Map, boolean)
+     */
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        super.configure(configs, isKey);
+        
+        Object ve = configs.get(JsonSchemaSerDeConstants.REGISTRY_JSON_SCHEMA_VALIDATION_ENABLED);
+        this.validationEnabled = Utils.isTrue(ve);
+    }
+    
+    /**
+     * @see org.apache.kafka.common.serialization.Serializer#serialize(java.lang.String, java.lang.Object)
+     */
+    @Override
+    public byte[] serialize(String topic, T data) {
+        // Headers are required when sending data using this serdes impl
+        throw new UnsupportedOperationException();
+    }
+    
+    /**
+     * @see org.apache.kafka.common.serialization.Serializer#serialize(java.lang.String, org.apache.kafka.common.header.Headers, java.lang.Object)
+     */
+    @Override
+    public byte[] serialize(String topic, Headers headers, T data) {
+        if (data == null) {
+            return new byte[0];
+        }
+
+        // Now serialize the data
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            JsonGenerator generator = mapper.getFactory().createGenerator(baos);
+            if (validationEnabled) {
+                String artifactId = getArtifactId(topic, data);
+                Long globalId = getArtifactVersionGlobalId(artifactId, topic, data);
+                addSchemaHeaders(headers, artifactId, globalId);
+
+                SchemaValidator schemaValidator = schemaCache.getSchema(globalId);
+                generator = api.decorateJsonGenerator(schemaValidator, generator);
+            }
+            addTypeHeaders(headers, data);
+
+            mapper.writeValue(generator, data);
+            
+            return baos.toByteArray();
+        } catch (IOException e) {
+            throw new SerializationException(e);
+        }
+    }
+
+    /**
+     * Figure out the artifact ID from the topic name and data.
+     * @param topic
+     * @param data
+     */
+    protected String getArtifactId(String topic, T data) {
+        // Note - for JSON Schema, we don't yet have the schema so we pass null to the strategy.
+        return getArtifactIdStrategy().artifactId(topic, isKey(), null);
+    }
+
+    /**
+     * Gets the global id of the schema to use for validation.
+     * @param artifactId
+     * @param topic
+     * @param data
+     */
+    protected Long getArtifactVersionGlobalId(String artifactId, String topic, T data) {
+        if (getGlobalIdStrategy() == null) {
+            return latestVersionStrategy.findId(getClient(), artifactId, ArtifactType.JSON, null);
```

Comment:
Why not really just use existing globalIdStrategy instance? I still don't why the 2?

Reply:
It was supposed to just be a default.  But maybe you're right and it's not needed.  Just fail if it's not set...  Originally when I was using ArtifactId and Version this made more sense, but maybe not anymore.  I've changed this to make the intention more clear, but let me know if you think we should just fail when the strategy is not set.